### PR TITLE
drivers: flash: Fix nrf QSPI NOR driver to behave like flash

### PR DESCRIPTION
**Pull request by sigvartmh**
Add the possibility to use heap to emulate internal flash driver more
closely.

Signed-off-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

### CODE REVIEW

File: drivers/flash/nrf_qspi_nor.c [hunk 1]
```diff
@@ -499,35 +528,82 @@ static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
...
+		memcpy(tx_ram_buf, src, var_size);
+		src = tx_ram_buf;
+	}
+#endif
+
 	/* write size must be multiple of 4 bytes */
 	if (size % sizeof(uint32_t) ||
 	    !(size > 0)) {
+#if (CONFIG_HEAP_MEM_POOL_SIZE > 0)
+		ret = -EINVAL;
```
Comment (hakonfam) [inline drivers/flash/nrf_qspi_nor.c]: See if you can do `goto out` in both these cases, then you can check the `if CONFIG_HEA_MEM_POOL_SIZE` inside the goto. This way the code flow won't be that different in the two different cases.

Comment (hakonfam) [reply drivers/flash/nrf_qspi_nor.c]: This applies to this whole patch.

**Review by hakonfam — CHANGES_REQUESTED** [review]

File: drivers/flash/nrf_qspi_nor.c (hunk 1 above)
Comment (sigvartmh) [reply drivers/flash/nrf_qspi_nor.c]: Forgot to push the updated version to upstream fork :) Fixed.

**Comment by nvlsianpu** [comment]
@sigvartmh  please elaborate what you mean by:
> Add the possibility to use heap to emulate internal flash driver more closely.

File: drivers/flash/nrf_qspi_nor.c [hunk 2]
```diff
@@ -476,17 +476,42 @@ static inline int qspi_nor_read_id(struct device *dev,
...
+	rx_ram_buf = NULL;
+	void *tmp_dest = dest;
+	size_t var_size = size;
+
+	if (((uint32_t)dest < CONFIG_SRAM_BASE_ADDRESS)
+	   || (size < sizeof(uint32_t))) {
+		if (size < sizeof(uint32_t)) {
+			size = 4;
+		}
+		rx_ram_buf = k_malloc(size);
```
Comment (nvlsianpu) [inline drivers/flash/nrf_qspi_nor.c]: dosen't malloc for 4 B is a overkill here?
Why no to use a variable on stack instead?

**Review by nvlsianpu — COMMENTED** [review]
I believe it would be much better to rework readout in order to have unaligned access without any constrains on HEAP. For instance nRF flash driver does it in three step:
1. Read 4-byte word which consist unaligned beginning (copy to caller buffer only requested data)
2. Read 4-byte aligned area
3. Read 4-byte word which consist unaligned ending (copy to caller buffer only requested data)
refer to https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/flash/soc_flash_nrf.c#L505-L563

File: drivers/flash/nrf_qspi_nor.c (hunk 2 above)
Comment (sigvartmh) [reply drivers/flash/nrf_qspi_nor.c]: I was unsure if the stack lived throughout the read. As I assumed it was DMA based. If it does then sure

**Comment by sigvartmh** [comment]
@nvlsianpu  Agree with your points I'll rework it so that it uses a similar technique as the flash driver. What I mean is basically add an option which does what `CONFIG_SOC_FLASH_NRF_EMULATE_ONE_BYTE_WRITE_ACCESS` does.

**Comment by nvlsianpu** [comment]
@sigvartmh ^^ Great.

**Comment by sigvartmh** [comment]
Closing this one couldn't find a good way to do it. It will have to wait a bit.

**Pull request closed without merging**